src/scripts/utils/tokens.ts

```diff
@@ -1,63 +1,95 @@
-  decimals: number
-  name: string
+  symbol?: string
+  decimals?: number
+  name?: string
+}
+interface Call {
+  target: string
+  callData: string
 }
 
-export async function getTokenInfos(tokenAddresses: string[], _multicall: Contract, provider: JsonRpcProvider): Promise<Map<string, Erc20Info>> {
+
+export async function getTokenInfos(tokenAddresses: string[], multicall: Contract, provider: JsonRpcProvider): Promise<Map<string, Erc20Info>> {
   const tokenInfos = new Map<string, Erc20Info>();
-  let i = 0
-  for (const address of tokenAddresses) {
-    console.log('Get token info for ', address, i++)
-    const erc20Contract = new Contract(address, ERC20_ABI, provider)
-    const symbol = await erc20Contract.symbol().catch(handleErrorReturnEmptyString)
-    const decimals = await erc20Contract.decimals().catch((e: any) => {
-      console.error('Error getting decimals for ', address, e)
-      return 18
+
+  if (tokenAddresses.length === 0) {
+    return tokenInfos
+  }
+
+  const erc20Contract = new Contract(tokenAddresses[0], ERC20_ABI, provider)
+
+  const { symbolCalls, nameCalls, decimalsCalls } = tokenAddresses.reduce<{ symbolCalls: Call[], nameCalls: Call[], decimalsCalls: Call[] }>((acc, address) => { // .slice(0, 2)
+
+    acc.symbolCalls.push({
+      target: address,
+      callData: erc20Contract.interface.encodeFunctionData('symbol')
+    })
+    acc.nameCalls.push({
+      target: address,
+      callData: erc20Contract.interface.encodeFunctionData('name')
     })
-    const name = await erc20Contract.name().catch(handleErrorReturnEmptyString)
+    acc.decimalsCalls.push({
+      target: address,
+      callData: erc20Contract.interface.encodeFunctionData('decimals')
+    })
+
+    return acc
+  }, {
+    symbolCalls: [],
+    nameCalls: [],
+    decimalsCalls: []
+  });
+
 
-    tokenInfos.set(address.toLowerCase(), {
+  // Make 3 multicalls, one for symbols, one for names and one for decimals
+  const [symbolsResult, namesResult, decimalsResults] = await Promise.all([
+    multicall.callStatic.tryAggregate(false, symbolCalls),
+    multicall.callStatic.tryAggregate(false, nameCalls),
+    multicall.callStatic.tryAggregate(false, decimalsCalls)
+  ]) as [ReturnCall[], ReturnCall[], ReturnCall[]]
+
+  for (let i = 0; i < symbolsResult.length; i++) {
+    const address = tokenAddresses[i]
+
+    const { success: successSymbol, returnData: symbolData } = symbolsResult[i];
+    const { success: successName, returnData: nameData } = namesResult[i];
+    const { success: successDecimals, returnData: decimalsData } = decimalsResults[i];
+
+    const symbol = successSymbol ? erc20Contract.interface.decodeFunctionResult('symbol', symbolData)[0] : undefined
+    const name = successName ? erc20Contract.interface.decodeFunctionResult('name', nameData)[0] : undefined
+    const decimals = successDecimals ? erc20Contract.interface.decodeFunctionResult('decimals', decimalsData)[0] : undefined
+
+
+    if (!symbol && !name && !decimals) {
```

Comment:
We should fail when any fail rather than all of them, no?

Reply:
Not necesarilly, those fields are not mandatory, this is why all these are optional in `Erc20Info` 

If there's none, there's no point on adding this to the result, because the merging of this info with the token will not add any information (this is why it skips the token)

Reply:
Yes, I know all of those methods are optional.
But I would argue, for the sake of completeness and the tidiness of the token list, we should not add a token that doesn't have all of those fields.

Reply:
It won't add it, it just ignores it if at the end of the information merge process ends up without those details. This is because the token list enforces these data to be present (otherwise it won't be compatible with Uniswap Token Lists and won't import in CoW Swap for example)

So I guess I don't need to change anything, right?